### PR TITLE
Add component search index service

### PR DESCRIPTION
**Pull request by Mbeaulne**
## Tophatting

No manual tophatting required. This PR only adds the local search index service and unit tests.

## What changed

Adds the local search index used by the new Components V2 page.

This is the plain text search layer. It turns hydrated component references into searchable entries and ranks matches by useful fields like name, description, inputs, outputs, and container command text.

## Why

The UI needs fast search before any AI reranking happens. This keeps normal search instant and predictable.

## Test plan

- Run `pnpm exec vitest run src/services/componentSearchIndex.test.ts`
- Run `pnpm run typecheck --pretty false`
- Sanity check that search ranks name matches above weaker matches like command-only matches

### CODE REVIEW

**Comment by Mbeaulne** [comment]
* **#2337** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2337?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#2350](https://github.com/TangleML/tangle-ui/pull/2350) <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2350?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#2351](https://github.com/TangleML/tangle-ui/pull/2351) <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2351?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#2323** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2323?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2322** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2322?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2319** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2319?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2318** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2318?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2317** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2317?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2321** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2321?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2320** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2320?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2316** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2316?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#2315** <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2315?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/TangleML/tangle-ui/2315?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `master`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Review by morgan-wowk — APPROVED** [review]

File: src/services/componentSearchIndex.ts [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+    for (const field of fields) {
+      if (entry.searchable[field].includes(token)) {
+        score += FIELD_WEIGHTS[field];
+        matched.add(field);
+      }
+    }
+  }
+
+  // Multi-token contiguous match in the name is a very strong signal.
+  if (tokens.length > 1 && entry.searchable.name.includes(fullQuery)) {
```
Comment (camielvs) [inline src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated code review comment.

**Full-substring bonus never fires for the snake_case case it documents.**

`fullQuery` keeps its original separators (spaces), while `searchable.name` keeps its original separators (underscores). So for the documented example — query `"train test split"` vs name `"train_test_split"` — `"train_test_split".includes("train test split")` is `false`, and the +10 bonus never applies. The doc comment just above explicitly claims this case is caught, and the test `"applies the multi-token full-substring bonus on the name"` passes only because per-token scoring already ranks `a` first — the bonus path is never exercised.

The bonus only fires for names containing literal spaces (e.g. `"Train Test Split"`), not snake_case — the most common component-name style here.

Suggested fix: normalize separators on both sides before the substring check, e.g.
```ts
const normalizedName = entry.searchable.name.replace(/[^a-z0-9]+/g, " ");
const normalizedQuery = tokens.join(" ");
if (tokens.length > 1 && normalizedName.includes(normalizedQuery)) {
  score += 10;
  matched.add("name");
}
```
Then strengthen the test so it would fail without the bonus.

**Comment by camielvs** [comment]
You may want to consider if we need such detailed comments

File: src/services/componentSearchIndex.ts [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+/** Which field of a component matched the query. Surfaced in the UI. */
+type MatchField = "name" | "description" | "io" | "implementation";
+
+/**
+ * Where a component came from. Attached to every index entry and threaded
+ * through to UI cards as a source badge so users know whether a result is
+ * from the curated standard library, the backend's published catalog, a
+ * registered external library (e.g. GitHub), or their own user components.
+ */
+export interface ComponentSource {
```
Comment (camielvs) [inline src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated code review comment.

**`ComponentSource` name collides with an existing, different concept.**

There is already a `ComponentSource` in `src/utils/componentTracking.ts` — a string union (`"user" | "library" | "published" | "url" | "file" | "unknown"`) describing component provenance for analytics. This adds a second, exported `ComponentSource` interface with overlapping `kind` values (`"user"`, `"published"`).

Two identically-named provenance types will confuse readers and auto-imports, and the `kind` unions will drift over time. Consider reusing/extending the existing union, or renaming this one to something more specific like `ComponentSearchSource` / `SearchResultSource`.

File: src/services/componentSearchIndex.ts [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+/**
+ * Split a query into lowercase alphanumeric tokens. `train_test_split` becomes
+ * `["train", "test", "split"]` — users almost always type the parts
+ * individually, and exact-string matches are still caught by substring search
+ * on the original lowercased text.
+ */
+function tokenize(text: string): string[] {
+  return text
+    .toLowerCase()
+    .split(/[^a-z0-9]+/i)
```
Comment (camielvs) [inline src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated code review comment.

**Redundant `i` flag on the tokenize regex.**

The input is already lowercased on the prior line, so the `/i` flag on `/[^a-z0-9]+/i` is dead. Dropping it makes the intent clearer and matches the `[a-z0-9]` character class.

File: src/services/componentSearchIndex.ts (hunk 1 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated response.

Fixed — normalized separators on both sides before the substring check, exactly as suggested. Both `searchable.name` and the joined query now collapse `[^a-z0-9]+` runs to single spaces before comparison, so the bonus fires for `train_test_split` matching `"train test split"`.

Also strengthened `"applies the multi-token full-substring bonus on the name"` with two local fixtures that score identically on per-token weight (3*5=15 each) but where one has the query tokens scattered (`a_train_b_test_c_split_d`) and one has them contiguous (`train_test_split`). Without the +10 bonus the scattered name wins on the alphabetical tiebreak; with it the contiguous one pulls ahead. Verified the test fails when the fix is reverted.

File: src/services/componentSearchIndex.ts (hunk 2 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated response.

Fixed — renamed `ComponentSource` to `ComponentSearchSource` in `componentSearchIndex.ts` (and updated the test file imports) so it doesn't collide with the analytics provenance union in `src/utils/componentTracking.ts`. Confined the rename to this service since extending the existing union would also drag analytics-only kinds (`url`, `file`, `unknown`) into the search-result shape.

File: src/services/componentSearchIndex.ts (hunk 3 above)
Comment (Mbeaulne) [reply src/services/componentSearchIndex.ts]: > 🤖 This is an AI-generated response.

Fixed — dropped the `/i` flag on the tokenize regex.

**Comment by Mbeaulne** [comment]
### Merge activity

* **Jun 5, 2:37 PM UTC**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2315).
* **Jun 5, 2:38 PM UTC**: [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2315) rebased this pull request as part of a merge.
* **Jun 5, 2:41 PM UTC**: @Mbeaulne merged this pull request with [Graphite](https://app.graphite.com/github/pr/TangleML/tangle-ui/2315).
